crypto/params.c

```diff
@@ -0,0 +1,258 @@
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+
+static const OSSL_PARAM *OSSL_PARAM_locate(const OSSL_PARAM *p, const char *key)
+{
+    if (p != NULL && key != NULL)
+        for (; p->key != NULL; p++)
+            if (strcmp(key, p->key) == 0)
+                return p;
+    return NULL;
+}
+
+static int OSSL_PARAM_get_int_common(const OSSL_PARAM *p, const char *key,
+                                     void *val, size_t sz)
+{
+    int neg, sign;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL)
+        return 0;
+
+    sign = p->data_type == OSSL_PARAM_INTEGER;
+    if (!sign && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER)
+        return 0;
+    if (p->buffer_size == sz) {         /* Fast path */
+        memcpy(val, p->buffer, sz);
+    } else if (p->buffer_size < sz) {   /* Widening */
+        const size_t eb = sz - p->buffer_size;
+
+#ifdef L_ENDIAN
+        neg = sign && (((unsigned char *)p->buffer)[p->buffer_size - 1] & 0x80);
+        memcpy(val, p->buffer, p->buffer_size);
+        memset(((unsigned char *)val) + p->buffer_size, neg ? 0xff : 0, eb);
+#else
+        neg = sign && (*(unsigned char *)p->buffer & 0x80);
+        memset(val, neg ? 0xff : 0, eb);
+        memcpy(((unsigned char *)val) + eb, p->buffer, p->buffer_size);
+#endif
+    } else {                            /* Narrowing */
+#ifdef L_ENDIAN
+        memcpy(val, p->buffer, sz);
+#else
+        memcpy(val, ((unsigned char *)p->buffer) + (p->buffer_size - sz), sz);
+#endif
+    }
+    return 1;
+}
+
+static int OSSL_PARAM_set_int_common(const OSSL_PARAM *p, const char *key,
+                                     const void *val, size_t sz)
+{
+    int neg, sign;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL)
+        return 0;
+
+    sign = p->data_type == OSSL_PARAM_INTEGER;
+    if (!sign && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER)
+        return 0;
+    if (p->buffer_size == sz) {         /* Fast path */
+        memcpy(p->buffer, val, sz);
+    } else if (p->buffer_size > sz) {   /* Widening */
+        const size_t eb = p->buffer_size - sz;
+
+#ifdef L_ENDIAN
+        neg = sign && (((unsigned char *)val)[p->buffer_size - 1] & 0x80);
+        memcpy(p->buffer, val, sz);
+        memset(((unsigned char *)p->buffer) + sz, neg ? 0xff : 0, eb);
+#else
+        neg = sign && (*(unsigned char *)val & 0x80);
+        memset(p->buffer, neg ? 0xff : 0, eb);
+        memcpy(((unsigned char *)p->buffer) + eb, val, sz);
+#endif
+    } else {                            /* Narrowing */
+#ifdef L_ENDIAN
+        memcpy(p->buffer, val, p->buffer_size);
+#else
+        memcpy(p->buffer, ((unsigned char *)val) + (sz - p->buffer_size),
+               p->buffer_size);
+#endif
+    }
+    if (p->return_size != NULL)
+        *p->return_size = sz;
+    return 1;
+}
+
+#define PARAM_INT(name, type) \
+    int OSSL_PARAM_get_##name(const OSSL_PARAM *p, const char *key, type *val) \
+    { \
```

Comment:
For the "fast path" (i.e. exactly correct size), I think it would be better to have the code here, i.e.

``` C
        if (sizeof(type) == p->buffer_size)
            return (*val = *(type *)p->buffer), 1;
```

and leave only the widening / narrowing bits in `OSSL_PARAM_get_int_common`

Reply:
It will need a type check as well, but that's easy enough.

Reply:
Done, it needed the locate call too.